### PR TITLE
feat: add paid upgrade modal on reaching insight limits

### LINKED ISSUE #2710: Feature: Add paid upgrade banner and modal for large contributor and repository insights
### Type of feature

🍕 Feature

### Current behavior

We need a prompt for high usage accounts to upgrade. We can use this context clue as a banner for eligible insights.

image

And add some sort of modal to explain benefits of upgrading

image

cc: @isabensusan 

### Suggested solution

_No response_

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [ ] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by zeucapua**
## Description

Implements a modal that will nudge the user to upgrade their workspace if the current insight goes over the limits (>10 contributors, >100 repositories).
- [x] Repository Insights
- [x] Contributor Insights

## What type of PR is this? (check all applicable)

- [X] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Closes #2710 

## Mobile & Desktop Screenshots/Recordings
image

https://github.com/open-sauced/app/assets/20603494/ea457e7c-b101-49de-9f9a-654f0a11fe0e

## Steps to QA

Both repository and contributor insights, but the first has a limit of 100 so... 
1. Go to a non upgraded workspace's contributor insights
2. Add more than 10 contributors (initial and want to be added)
3. See upgrade modal, and click to upgrade
4. Be redirected to workspace settings and upgrade card

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [X] 🙅 no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?
i-phone-gif

### CODE REVIEW

**Comment by zeucapua** [comment]
@isabensusan any changes needed for the design?

File: components/Workspaces/InsightUpgradeModal.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                  <ChecklistItem color="green">Private or public Workspaces, you choose!</ChecklistItem>
+                  <ChecklistItem color="green">
+                    Up to <span className="font-bold">1000</span> contributors per insight page
+                  </ChecklistItem>
+                  <ChecklistItem color="green">
+                    Up to <span className="font-bold">10,000</span> repositories per insight page
+                  </ChecklistItem>
+                </ul>
+
+                <Button variant="primary" className="py-3 flex justify-center">
```
Comment (nickytonline) [inline components/Workspaces/InsightUpgradeModal.tsx:93]: Refactor (nit): if it's centering text only, prefer `text-center`

```suggestion
                <Button variant="primary" className="py-3 text-center">
```

File: components/Workspaces/InsightUpgradeModal.tsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                    Up to <span className="font-bold">10</span> contributors per insight page
+                  </ChecklistItem>
+                  <ChecklistItem color="orange">
+                    Up to <span className="font-bold">100</span> repositories per insight page
+                  </ChecklistItem>
+                </ul>
+
+                <Button
+                  variant="text"
+                  className="flex w-full py-3 justify-center !border !border-orange-600 text-orange-600"
```
Comment (nickytonline) [inline components/Workspaces/InsightUpgradeModal.tsx:62]: There currently is no variant `text`. Is that a new variant you have on your local? Is this a new variant we need @isabensusan? If we're introducing a new variant, are these the styles for that variant, `border border-orange-600 text-orange-600`?

**Comment by bdougie** [comment]
@isabensusan for the upgraded experience, we will want the numbers to be 

1000 repos and 100 contributors (unless we are confident we can display 1000 contributor lists without issue). 

But 10k repos seems like we an enterprise conversation.

**Comment by isabensusan** [comment]
@bdougie yep! I updated zeu on this after our call -- if you're referring to the mockup in [this issue](https://github.com/open-sauced/app/issues/2710), I just added a comment updating it

File: components/Workspaces/InsightUpgradeModal.tsx (hunk 1 above)
Comment (zeucapua) [reply components/Workspaces/InsightUpgradeModal.tsx]: `text-center` doesn't work, weirdly

**Comment by zeucapua** [comment]
Needs to be changed from blocking insight updates to a reminder after updating.

File: .env [hunk 3]
```diff
@@ -4,7 +4,7 @@ NEXT_PUBLIC_API_URL=https://beta.api.opensauced.pizza/v2
 NEXT_PUBLIC_EXP_API_URL=https://beta.api.opensauced.pizza/v2
 NEXT_PUBLIC_POSTHOG_ID=phc_Y0xz6nK55MEwWjobJsI2P8rsiomZJ6eZLoXehmMy9tt
 NEXT_PUBLIC_CLOUD_NAME=dgxgziswe
-NEXT_PUBLIC_STRIPE_PUBLISHABLE_KEY=test
+NEXT_PUBLIC_STRIPE_PUBLISHABLE_KEY=pk_test_51H1yp3JnZ7CYWRG2KmQMIE8sizoaSaZ8FxlEsc5uQb81gVswDlWu6FErmcukqc6Hso3mhCWgnLsjWphight4t1Mf00jZZjckyZ
```
Comment (brandonroberts) [inline .env:7]: Revert this change. You can create a `.env.local` file at the root of the project next to the `.env` and put your local environment variable changes there. Its not picked up by source control, but Next.JS will pick them up when running the app

**Comment by zeucapua** [comment]
Latest iteration: nudge banner on dashboard/overview of insight page, on click will open the modal

https://github.com/open-sauced/app/assets/20603494/99170bf1-9d4d-4c33-91ef-98d748e945c0

**Comment by zeucapua** [comment]
Changed to nudge banner on the dashboard if the insight page is over the limit.

https://github.com/open-sauced/app/assets/20603494/1b9df6c7-c1b8-45f6-851f-0365cc1437a6

File: components/organisms/InsightPage/InsightPage.tsx [hunk 4]
```diff
@@ -613,7 +618,13 @@ const InsightPage = ({ edit, insight, pageRepos, workspaceId }: InsightPageProps
           edit={edit}
           hasItems={repos.length > 0}
           handleCreatePage={handleCreateInsightPage}
-          handleUpdatePage={handleUpdateInsightPage}
+          handleUpdatePage={() => {
+            if (!isWorkspaceUpgraded && repos.length > 100) {
```
Comment (brandonroberts) [inline components/organisms/InsightPage/InsightPage.tsx:622]: Let's remove the blockers from updating

File: pages/workspaces/[workspaceId]/contributor-insights/[listId]/add-contributors.tsx [hunk 5]
```diff
@@ -211,8 +217,15 @@ const AddContributorsToList = ({ list, workspaceId, timezoneOption }: AddContrib
   const [contributors, setContributors] = useState<DbPRContributor[]>([]);
   const [suggestions, setSuggestions] = useState<string[]>([]);
   const [searchResults, setSearchResults] = useState<GhUser[]>([]);
+  const [isInsightUpgradeModalOpen, setIsInsightUpgradeModalOpen] = useState(false);
+  const { data: isWorkspaceUpgraded } = useIsWorkspaceUpgraded({ workspaceId: workspaceId! });
 
   const addContributorsToList = async () => {
+    if (!isWorkspaceUpgraded && initialCount + selectedContributors.length > 10) {
```
Comment (brandonroberts) [inline pages/workspaces/[workspaceId]/contributor-insights/[listId]/add-contributors.tsx:224]: Let's remove the blockers from updating

**Review by brandonroberts — APPROVED** [review]

**Review by jpmcb — APPROVED** [review]
